File: crates/taflow-core/src/indicators/mc_clellan_oscillator.rs
use crate::error::TaResult;
/// Fast-minus-slow EMA of ratio-adjusted net advances.
#[derive(Debug, Clone)]
pub struct McClellanOscillator {
    fast: Option<f64>,
    slow: Option<f64>,
    count: usize,
    value: Option<f64>,
}
impl McClellanOscillator {
    /// Create an unseeded classic 19/39 oscillator state.
    pub fn new() -> TaResult<Self> {
        Ok(Self {
            fast: None,
            slow: None,
            count: 0,
            value: None,
        })
    }
    /// Append aggregate advancing and declining issue counts.
    pub fn append(&mut self, advancers: f64, decliners: f64) -> Option<f64> {
        self.count += 1;
        let breadth = (advancers - decliners) / (advancers + decliners).max(1.0) * 1000.0;
        let f = self.fast.map_or(breadth, |x| x + 0.1 * (breadth - x));
        let s = self.slow.map_or(breadth, |x| x + 0.05 * (breadth - x));
        self.fast = Some(f);
        self.slow = Some(s);
        self.value = Some(f - s);
        self.value
    }
    /// Return the latest oscillator value.
    pub fn value(&self) -> Option<f64> {
        self.value
    }
    /// Return the number of processed market ticks.
    pub fn len(&self) -> usize {
        self.count
    }
    /// Return whether no market ticks have been processed.
    pub fn is_empty(&self) -> bool {
        self.count == 0
    }
    /// Clear both EMA seeds and the latest value.
    pub fn reset(&mut self) {
        self.fast = None;
        self.slow = None;
        self.count = 0;
        self.value = None;
    }
}

File: python/taflow/indicators/mc_clellan_oscillator.py
"""Python adapter for the native McClellan Oscillator."""

from typing import Any
import numpy as np
from .._native import McClellanOscillator as _Native
from .._series import as_float64_series


class McClellanOscillator:
    """Subtract slow from fast EMA of ratio-adjusted net advances.

    Ratio-adjusted net advances are ``1000 * (advancers - decliners) /
    max(advancers + decliners, 1)``. The native 19/39 EMA pair is seeded on the
    first tick, which emits zero. This maps to Wickra ``McClellanOscillator``.

    Args:
        advancers: Number of advancing issues at each tick.
        decliners: Number of declining issues at each aligned tick.

    Raises:
        ValueError: If the two histories have different lengths.
    """

    def __init__(self) -> None:
        """Initialize an empty configured native state.
        """
        self._state = _Native()

    def append(self, advancers: float, decliners: float) -> "McClellanOscillator":
        """Append one breadth tick and return this adapter."""
        self._state.append(float(advancers), float(decliners))
        return self

    def extend(self, advancers: Any, decliners: Any) -> "McClellanOscillator":
        """Append aligned breadth histories after length validation."""
        arrays = as_float64_series(advancers), as_float64_series(decliners)
        if len(arrays[0]) != len(arrays[1]):
            raise ValueError("advancers and decliners must have equal lengths")
        self._state.extend(*arrays)
        return self

    @property
    def value(self) -> float | None:
        """Return the latest oscillator value, or ``None`` before input."""
        return self._state.value

    def compute(self) -> np.ndarray:
        """Return the aligned oscillator history."""
        return self._state.compute()

    def reset(self) -> "McClellanOscillator":
        """Reset both EMA states and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the processed-tick count delegated to native state."""
        return len(self._state)
